### PR TITLE
Add golang-lint to the config pod CI and fix all lint warnings

### PR DESCRIPTION
**Pull request by msteffen**
This was precipitated by https://pachyderm.slack.com/archives/CSELV66P5/p1640124036326000, where the dropped error in `steps.go` caused an invalid config to fail silently, leading to some user confusion.

This PR fixes the dropped error, adds some basic linting that could catch that class of errors, and then fixes all the other lint errors (which are all in tests).

Other changes in this PR:
- Upgrade the pach dependency to 2.0.5 (unrelated to the other changes, but small :grimacing:)
- return `errSkipped` if an IDP config update is a no-op because the config isn't changed. Also, report skipped steps in the logs.

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -102,7 +101,9 @@ func main() {
 				stepLogger.WithError(err).Error("error syncing cluster state")
 				os.Exit(1)
 			}
-			stepLogger.WithField("reason", err).Info("skipped")
+			stepLogger.WithField("reason", err).Warn("skipped")
+		} else {
+			stepLogger.Info("success")
```
Comment (acohen4) [inline main.go:106]: 👍

File: steps.go [hunk 2]
```diff
@@ -172,7 +174,12 @@ func idpsStep(_ *client.APIClient, ec *client.APIClient) error {
 	}
 
 	for _, connector := range connectors {
-		updateOrCreateIDP(ec, connector, existing.Connectors)
+		if err := updateOrCreateIDP(ec, connector, existing.Connectors); err != nil {
```
Comment (acohen4) [inline steps.go:177]: good catch!

File: steps_test.go [hunk 3]
```diff
@@ -116,7 +116,7 @@ func (s *StepTestSuite) TestSimpleConfig() {
 	s.writeSimpleConfig()
 
 	for _, step := range syncSteps {
-		step.fn(s.c, s.c)
+		s.Require().NoError(step.fn(s.c, s.c))
```
Comment (acohen4) [inline steps_test.go]: I think these are failing in CI on the success case where we return `errSkipped`.

File: .circleci/config.yml [hunk 4]
```diff
@@ -1,6 +1,30 @@
...
+    steps: # steps that comprise the `build` job
+      - checkout # check out source code to working directory
+      - run: "echo $(($(date +%s)/604800)) > current_week"
+      - restore_cache:
+          key: golang-lint-ci-{{ checksum "current_week" }}
+      - run:
+          name: install lint
+          command: |
+            if [[ ! -e /usr/local/bin/golangci-lint ]]; then
+              cd /usr/local
```
Comment (acohen4) [inline .circleci/config.yml]: not sure what's causing the permissions issue that's failing in CI. Is there are a reason the docker image for this job is `circleci/golang:1.16`. In most of our CI we use `ubuntu-2004:202101-01`

Comment (msteffen) [reply .circleci/config.yml]: Probably just need to add `chmod +x` 🤦‍♀️

Comment (chainlink) [reply .circleci/config.yml]: Looks like there is a golangci docker image you could use https://hub.docker.com/r/golangci/golangci-lint

File: steps_test.go (hunk 3 above)
Comment (msteffen) [reply steps_test.go]: :+1: thanks! I added a helper function, `RequireNilOrSkipped` for these `step.fn` calls, which will hopefully fix these tests.

File: .circleci/config.yml (hunk 4 above)
Comment (msteffen) [reply .circleci/config.yml]: :+1: thanks! I switched to that image and removed the install step, which cleaned up `circleci/config.yml` a lot

**Review by acohen4 — APPROVED** [review]
LGTM!

File: .circleci/config.yml [hunk 5]
```diff
@@ -28,6 +37,7 @@ jobs:
 workflows:
   circleci:
     jobs:
+      - lint
```
Comment (chainlink) [inline .circleci/config.yml:40]: We should add
```
filters:
  tags:
    only: /^\d+\.\d+$/
```
So this runs on releases as well (Counterintuitively, this means that circle will run this on tags, as by default it does not)

File: .circleci/config.yml [hunk 6]
```diff
@@ -28,6 +37,7 @@ jobs:
...
       - test:
           filters:
             tags:
```
Comment (chainlink) [inline .circleci/config.yml:43]: We should require lint on release as well (line #47, but strangely I can't add a note to that line specifically.

**Review by chainlink — APPROVED** [review]
Could small changes, conditionally approving

**Comment by msteffen** [comment]
> Could small changes, conditionally approving

Whoops, sorry, I jumped the gun! I'll send a followup PR

**Update** followup is: https://github.com/pachyderm/config-pod/pull/8

**Pull request merged**